### ISSUE TITLE
Support by-ref-like (`ref struct`) parameter types such as `Span<T>` and `ReadOnlySpan<T>`

### ISSUE THREAD

**Issue by stakx** [castleproject/Core#663]
DynamicProxy does not currently support those because by-ref-like values cannot be boxed to be put in the `object[] IInvocation.Arguments` array.

I haven't yet been able to think of a general way how _all_ by-ref-like types (including user-defined ones) because of their various limitations. If anyone has ideas, I'd be interested in hearing them!

However, it might be fairly easy to at least add support for `Span<T>` and `ReadOnlySpan<T>` specifically. Those types are becoming more and more common in the .NET FCL, so while not an ideal solution, it might still be sufficient for most use cases:

We could introduce a new type in DynamicProxy's namespace:

```csharp
public interface ISpanMarshaller
{
    object BoxSpan<T>(Span<T> span);
    object BoxReadOnlySpan<T>(ReadOnlySpan<T> readOnlySpan);
    ReadOnlySpan<T> UnboxReadOnlySpan<T>(object boxedReadOnlySpan);
    Span<T> UnboxSpan<T>(object boxedSpan);
}
```

Then we could introduce a new property to `ProxyGenerationOptions`, `ISpanMarshaller SpanMarshaller { get; set; }`, which generated proxies could use to transfer spans into and out of the `object[] IInvocation.Arguments` array.

I don't really like an addition that deals with two very specific types, but they are becoming more and more common, and I haven't yet thought of anything better.

Opinions, or alternate ideas, anyone?

**Comment by stakx** [comment]
I've done some quick prototyping. The extension proposed above is feasible and works.

(I'd however perhaps make one change to the proposal: instead of adding `ISpanMarshaller SpanMarshaller { get; set; }` to `ProxyGenerationOptions`, it might be better to add a `Type SpanMarshallerType { get; set; }` instead. Proxies would then instantiate a marshaller instance themselves, instead of getting one injected. This has a few benefits:

 * Marshaller types do not need to be serializable.
 * It's clearer that they are supposed to be stateless (because it's out of the user's control when & how often they get instantiated).
 * It's easier to implement. :wink:)

**Comment by stakx** [comment]
Scratch the above proposal, I think I can actually come up with one that works for arbitrary by-ref-like types. I'll post an updated feature proposal shortly.

**Comment by stakx** [comment]
I've put something together that appears to work... see the draft PR linked above for two short code examples how this would work. I'd be glad for some feedback (and whether the suggested API would actually be usable e. g. in downstream testing libraries).

/cc @thomaslevesque & @blairconrad (FakeItEasy), @dtchepak (NSubstitute), @jonorossi

**Comment by thomaslevesque** [comment]
Hey @stakx, nice job!
Using converters is an interesting approach. I think it would work just fine for mocking scenarios, where performance isn't typically a critical concern. In other scenarios, the performance impact of converting spans to arrays is something to keep in mind. I don't see a better way, though...
Something else to consider: there might not always be a sensible way of converting a `ref struct` to an object and still be able to convert back (I'm thinking of `Utf8JsonReader`, for instance).
From FakeItEasy's perspective, I think the converters for `Span<T>`/`ReadOnlySpan<T>` would be provided directly by the library, and we could expose extension points to provide additional converters.
BTW, is there a particular reason why converters don't implement an interface in your proposal?

**Comment by stakx** [comment]
Thanks for the feedback @thomaslevesque.

> is there a particular reason why converters don't implement an interface in your proposal?

Yes, there is. `ref struct` types cannot be used as generic type arguments. So a proper interface type such as `IByRefLikeConverter<TByRefLike>` could never be instantiated.

**Comment by stakx** [comment]
> there might not always be a sensible way of converting a ref struct to an object and still be able to convert back

True. I don't see what we could do about that, though... do you? If I understand your concern correctly, this would make it only impossible to do round-tripping of `ref struct` values during interception. "Half-trips" should still work since the `Box` / `Unbox` methods are only looked up & called when needed; in theory, if you don't need the conversion back from `object` (i.e. your `ref struct` type does not figure as a method return type or in a `ref`/`out` parameter) you wouldn't even need to provide the `Unbox` conversion method.

**Comment by stakx** [comment]
> the performance impact of converting spans to arrays is something to keep in mind.

Indeed. Which is why the copying behavior shouldn't be the default behavior (for spans), and why I thought it would be a good idea for the converter selectors to pinpoint single method parameters, so you're free to choose to nullify most spans except where retaining the value somehow &ndash; e.g. as an array copy &ndash; is actually important. Being only able to set up a single converter globally that gets used for all by-ref-like parameters equally might be too coarse-grained.

I've also briefly considered doing weird `unsafe` pointer-based stuff (similar to what `System.Runtime.CompilerServices.Unsafe` does) instead of converters that simply convert to/from `object`,  but apart from being unsure if it would've even worked, I don't think it would've resulted in a very nice, easy to understand API).

**Comment by blairconrad** [comment]
Thanks, @stakx. Slightly over my head, but in general, I think I understand the approach and the tradeoffs. Appreciate your work.

The possible performance optimization by specifying the position of the parameter is interesting, and not something I would've thought of.

Just to make sure I understand everything, if I combine that power with a comment @thomaslevesque made earlier ("From FakeItEasy's perspective, I think the converters for `Span<T>/ReadOnlySpan<T>` would be provided directly by the library"), then I expect FakeItEasy would end up providing a selector not unlike the sample one you had:

```c#
// this will get invoked during proxy type generation:
class ConverterSelector : IByRefLikeConverterSelector
{
    public Type SelectConverterType(MethodInfo method, int parameterPosition, Type parameterType)
    {
        return parameterType == typeof(Span<byte>) ? typeof(CopyByteSpanConverter) : null;
    }
}
```

where we'd end up ignoring the `parameterPosition`.

**Comment by stakx** [comment]
@blairconrad, I suppose that you'd typically configure a converter out-of-the-box that checks for both `Span<*>` _and_ `ReadOnlySpan<*>`, and optionally lets user code opt into the conversion process for all other types. And yes, it should typically suffice to only check the type and ignore the `MethodInfo` and the parameter position.

For mocking libraries, the selector's method body could look like this:

```csharp
if (parameterType.IsGenericType)
{
    var def = parameterType.GetGenericTypeDefinition();
    var args = paramrterType.GetGenericArguments();
    if (def == typeof(Span<>))
    {
        return typeof(CopySpanConverter<>).MakeGenericType(args);
    }
    else if (def == typeof(ReadOnlySpan<>)
    {
        // as above but with a CopyReadOnlySpanConverter<> instead
    }
    else
    {
        // delegate to user-provided selector function,
        // or return null to let DynamicProxy nullify everything
    }
}
```

(Btw., I'm not 100% sure yet whether the `Box` method's parameter list is going to stay that way; it might also be possible to pass in a `ParameterInfo` which would tie those three parameters nicely together.)

**Comment by thomaslevesque** [comment]
> Yes, there is. `ref struct` types cannot be used as generic type arguments. So a proper interface type such as `IByRefLikeConverter<TByRefLike>` could never be instantiated.

Ah, yes, of course. Thanks for the reply!

> True. I don't see what we could do about that, though... do you?

No... I'm not even sure there _is_ a solution. I'm just pointing out a limitation.

> If I understand your concern correctly, this would make it only impossible to do round-tripping of `ref struct` values during interception. "Half-trips" should still work since the `Box` / `Unbox` methods are only looked up & called when needed; in theory, if you don't need the conversion back from `object` (i.e. your `ref struct` type does not figure as a method return type or in a `ref`/`out` parameter) you wouldn't even need to provide the `Unbox` conversion method.

Yeah, half-trips would probably work in most cases.

**Comment by stakx** [comment]
@zvirja, thanks for double-checking. I'm well aware that my pointer-based experiments are running dangerously close to unsafe territory, so I am glad for any input beyond my own thoughts.

I need to double check what your POC is really doing, but I think it is going wrong there:

> ```csharp
> var spanPointer = Unsafe.AsPointer(ref MemoryMarshal.GetReference(arg));
> ```

This does not take the address (in the form of an unmanaged pointer) of `arg` (the span) itself, but of the first element referenced by that span. In other word, it performs a double-dereference.

I am not planning on doing that, however. I would never be accessing (looking into) the byref-like arguments beyond taking their address. It is my understanding that while the GC may _update_ things that live on the evaluation stack (e. g. references to objects that it has moved to a new address), it will never _move_ things on the stack. And since byref-like values like `Span<>` can _only_ live on the stack, taking their addresses should be perfectly safe, even when GC happens once or multiple times.

Have I misunderstood either something fundamental, or what your POC is trying to show?

**Comment by zvirja** [comment]
Yes, you are definitely right in the way how you assess my code. I am indeed taking the address of the underlying object which is ultimately unsafe. Glad we both agree on that part.

Your idea about taking the reference itself is an interesting point of thought. In a very simplistic model that would work indeed, as if you imagine reference itself being stored on a stack, then indirectly passing address on stack shall be safe indeed while calling child functions. And then using de-referencing it underneath.

While you idea might work, it's not obvious to me, as we entering territory of ABI and how things are translated on the machine level. The concept "things live on stack" is a bit simplistic, as values could also live in registers and compiler could do all sorts of optimizations. Back in days on x86 32-bits they were inlining struct arguments directly into the registers under certain conditions, even when it spans more than 32 bit. For x86 64-bits they don't do it as aggressively anymore, but they definitely inline structs up to 64 bits. While all of this might not apply here (as `Span` is larger), I just want to highlight an idea that we are entering a low-level world and such thing as "address on stack" in certain cases might never exist in physical memory, as values will be in registers only.

The ideal solution for us would be use the abstractions that are promised by the language and will not break when you shift to another architecture. We don't want to depend on JIT specifics. Maybe if we play somehow with `stackalloc` it is doable, as there you indeed have address on stack. But last time I checked that it had limitations on primitive types only, so it might be impossible to marry both words.

If you find out a creative solution, please share - would be happy to review. I suggest starting with C# POC, so it's easier to play with ideas 😉 

We could also probably invite somebody from MS ask for there suggestions. We doing some black magic, but we have a real problem to solve. As we have more and more spans, it's such a pity that we cannot represent it anymore.

Another idea could be to just copy `Span/ReadOnlySpan` arg to an array, pass the value as array to interceptors and then copy back (for `Span`, but not `ReadOnlySpan`). It will be ultimately safe. Sure, there will be a small performance penalty - but it's not that bad. This way we don't support generic `ref struct`, but instead support `Span` and `ReadOnlySpan` specifically (and maybe other types). I like it because we play safe and it will work in all the architectures.

**Comment by stakx** [comment]
@zvirja, thanks for offering to review and helping to think this through a bit. I'll happily show some code. FYI, my latest draft lives here, in case you want to try it out: https://github.com/castleproject/Core/compare/master...stakx:Castle.Core:byref-like-arguments. I'll try to summarize its intention in a somewhat simplified form below.

Before I dump some code on you, just one remark regarding "black magic": I don't think it needs to be perceived as such. I am mostly looking at this problem space through the lens of ECMA-335 (the spec for the CLR and CoreCLR), which gives us an machine & execution model with certain features and guarantees (e. g. that there are typed locations such as method parameters and local variables that can hold a value; that one may take the address of such locations; that such addresses can be loaded onto the evaluation stack; and that certain instructions can take addresses both in the form of managed _or_ unmanaged pointers). I expect that the (Core)CLR's JIT adheres to that spec, too; therefore we don't necessarily need to think about implementation details such as register allocation or code inlining... but whatever we do _must_ adhere to ECMA-335.

---

Let me start with some code. First, let's forget about `System.Reflection.Pointer` and introduce our own wrapper types for unmanaged pointers. This allows us to give that type a reasonably type-safe surface. Say, for .NET 9+, we could have a wrapper type like this one:

```csharp
public unsafe class ByRefLikeWrapper<TByRefLike> where TByRefLike : struct, allows ref struct
{
    protected readonly void* ptr;

    [CLSCompliant(false)]
    [EditorBrowsable(EditorBrowsableState.Never)]
    public ByRefLikeWrapper(void *ptr)
    {
        this.ptr = ptr;
    }

    public ref TByRefLike Get()
    {
        return ref Unsafe.AsRef<TByRefLike>(this.ptr);
    }
}
```

(In my latest draft, I have a small [hierarchy of wrapper types](https://github.com/castleproject/Core/commit/a01b85f4361655e408f170bd92e9938c1075d478) because `allows ref struct` only works on .NET 9+.)

#### Byref-like arguments that get passed by value

Then say you're asking DynamicProxy to proxy a type having a method as follows:

```csharp
public interface IFoo
{
    void M(Span<char> characters);
}
```

DynamicProxy might then choose to implement this method as follows:

```csharp
partial class IFooProxy
{
    public virtual void M(Span<char> characters)
    {
        new InheritanceInvocationWithoutTarget(
            ...,
            arguments: [ new ByRefLikeWrapper<Span<char>>(Unsafe.AsPointer(ref characters) ])
        .Proceed();
    }
}
```

#### Byref-like arguments that are passed by reference (`in`, `ref`, `ref readonly`, or `out`)

```csharp
public interface IFoo
{
    void M(ref Span<char> characters);
}
```

DynamicProxy generally buffers those in local variables, then (in case of `ref` and `out`) copies those buffered values back at the end. The emitted code would then look roughly as follows (from memory):

```csharp
partial class IFooProxy
{
    public virtual void M(ref Span<char> characters)
    {
        Span<char> charactersCopy = characters;
        try
        {
            new InheritanceInvocationWithoutTarget(
                ...
                arguments: [ new ByRefLikeWrapper<Span<char>>(Unsafe.AsPointer(ref charactersCopy) ])
            .Proceed();
        }
        finally
        {
            characters = charactersCopy;
        }
    }
}
```

#### Methods with a byref-like return parameter

These would be a tad more difficult, and not make for a particularly self-explanatory or elegant API (IMHO):

```csharp
public interface IFoo
{
    Span<char> M();
}
```

Since the byref-like wrapper type has to point at a valid storage location, and the method interceptors cannot easily come up with such a one that is unique per call, and one that lives beyond the `Intercept` call, the proxied method body would likely have to provide a return value storage location for the interceptors:

```csharp
partial class IFooProxy
{
    public virtual Span<char> M()
    {
        Span<char> returnValueBuffer = default;
        var returnValueWrapper = new ByRefLikeWrapper<Span<char>>(Unsafe.AsPointer(ref returnValueBuffer));
        new InheritanceInvocationWithoutTarget(...)
        {
            ReturnValue = returnValueWrapper,
        }
        .Proceed();
        return returnValueWrapper.Get();
    }
}
```

Interceptors would then be able to return a byref-like value as follows:

```csharp
partial class InterceptorForM : IInterceptor
{
    public void Intercept(IInvocation invocation)
    {
        ref Span<char> returnValue = ref ((ByRefLikeWrapper<Span<char>>)invocation.ReturnValue).Get();
        returnValue = ...;
    }
}
```

Or in case where the intercepted method also has a byref-like parameter of the same type, one could just copy the whole wrapper: `invocation.ReturnValue = invocation.Arguments[n]` and the proxied method body would fetch the value referenced by it.

Not saying it's a particularly self-explanatory or elegant API, but the best one I could come up so far. It may also be feasible to implement C# `ref` returns that way.)

---

#### Some closing remarks (for today):

* Any such wrapper objects would have to be invalidated at the end of interception, in case someone tries to copy them away and use them later. That would be an error since the method arguments (or local variables from the proxied method body) would be gone from the stack by then. My current draft has a commit adding such invalidation logic (though it's not optimized yet).

* I've also considered a whole other approach where end users never get to see a wrapper type like `ByRefLikeWrapper<TByRef>`. Instead, they'd get new generic method overloads for `invocation.GetArgumentValue<T>`, `invocation.SetArgumentValue<T>`, `invocation.GetReturnValue<T>` and `invocation.SetReturnValue<T>` that would make use of .NET 9+ `where T : allows ref struct` anti-constraints. This might be the nicer approach but suffers from one issue, namely that if one by-passes these methods and uses the `invocation.Arguments` array or `invocation.ReturnValue` directly, one would probably just see `null` placeholders for any byref-like values. Not sure how well that would work for mocking libraries which (I suspect) probably prefer these properties over the corresponding methods.

**Comment by zvirja** [comment]
Hey @stakx!

Sorry for a bit belated reply. I give it a bit of thought and played with it for a bit and I found a couple of issues. I also have suggestions how to make it totally safe.

First of all I would say kudos for the impressive bit of work! It's a brilliant idea and I love the implementation of it, even though I don't follow all the internals. I was not skeptical about it since the long time ago you explained it and I since then I was totally with you 😉 Thank you also for the paranoid mode with preserving stack pointers 😅 

I used the following test program to play with it:

<details><summary>Code</summary>

```c#
#define ORIGINAL_API

using Castle.DynamicProxy;

static class Program
{
    internal static void Main(string[] args)
    {
        var proxyGenerator = new ProxyGenerator();
        var proxy = proxyGenerator.CreateInterfaceProxyWithoutTarget<ISomething>(new PlaygroundInterceptor());

        var result1 = InvokeMethod1(proxy);
        Console.WriteLine($"Method 1 OUTER. Result: {result1.PrintSpan()}");
        new List<int> { 1, 2, 3 }.Reverse();
        Console.WriteLine($"Method 1 OUTER. Result: {result1.PrintSpan()}");
        
        InvokeMethod2(proxy, out ReadOnlySpan<int> result2);
        Console.WriteLine($"Method 2 OUTER. Result: {result2.PrintSpan()}");
        
        Console.ReadLine();
    }

    private static Span<int> InvokeMethod1(ISomething something)
    {
        Span<int> values = stackalloc int[] { 11, 22 };

        var result = something.Method1(values);
        Console.WriteLine($"Method 1. Result: {result.PrintSpan()}");

        return result;
    }

    private static void InvokeMethod2(ISomething something, out ReadOnlySpan<int> result)
    {
        Span<int> values = stackalloc int[] { 11, 22 };

        result = something.Method2(values);
        Console.WriteLine($"Method 2. Result: {result.PrintSpan()}");
    }
}

public class PlaygroundInterceptor : IInterceptor
{
    public void Intercept(IInvocation invocation)
    {
        switch (invocation.Method.Name)
        {
            case "Method1":
                HandleMethod1(invocation);
                break;
            
            case "Method2":
                HandleMethod2(invocation);
                break;
            
            default:
                throw new NotImplementedException();
        }
    }

    private void HandleMethod1(IInvocation invocation)
    {
        var arg1 = (SpanReference<int>)invocation.Arguments[0]!;
        var returnValue = (SpanReference<int>)invocation.ReturnValue!;

        var mre = new ManualResetEventSlim(false);
        
        Task.Run(() => ConcurrentInvocationHandler());

        mre.Wait();

#if ORIGINAL_API
        returnValue.Value = arg1.Value;
#else
        returnValue.SetValue(arg1.UseValue((scoped value) => value.ToArray().AsSpan()));
#endif
        
        void ConcurrentInvocationHandler()
        {
            try
            {
                var arg1 = (SpanReference<int>)invocation.Arguments[0]!;
#if ORIGINAL_API
                var arg1Span = arg1.Value;
                
                mre.Set();
                Thread.Sleep(TimeSpan.FromMilliseconds(500));
#else
                var arg1Span = arg1.UseValue((scoped x) =>
                {
                    mre.Set();
                    Thread.Sleep(TimeSpan.FromMilliseconds(500));    
                    
                    return x.ToArray().AsSpan();
                });
#endif

                Console.WriteLine($"Method 1 Concurrent. Arg1 value: {arg1Span.PrintSpan()}");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Method 1 Concurrent. Error occurred: {ex}");
            }
        }
    }

    private void HandleMethod2(IInvocation invocation)
    {
        var arg1 = (SpanReference<int>)invocation.Arguments[0]!;
        var returnValue = (SpanReference<int>)invocation.Arguments[0]!;

#if ORIGINAL_API
        returnValue.Value = arg1.Value;
#else
        returnValue.SetValue(arg1.UseValue((scoped value) => value.ToArray().AsSpan()));
#endif
    }
}

public interface ISomething
{
    Span<int> Method1(scoped Span<int> arg1);
    
    Span<int> Method2(scoped Span<int> arg1);
}

public static class SpanExtensions
{
    extension<T>(ReadOnlySpan<T> span)
    {
        public string PrintSpan()
        {
            return string.Join(", ", span.ToArray().Select(x => x?.ToString()));
        }
    }
} 

```
</details> 

Output I got on my machine:

```
Method 1. Result: 11, 22
Method 1 OUTER. Result: 11, 22
Method 1 OUTER. Result: -639832728, 79
Method 2. Result: 
Method 2 OUTER. Result: 
Method 1 Concurrent. Arg1 value: 0, 0
```

## Issues

### Issue 1
I tried to make stack-allocated value to escape and I succeeded by using a couple of techniques:
- I used concurrency
- I bypassed the `scoped` constrain and managed to return the value via `return`/`ref/out args`.

As you can see, `Method 1 OUTER. Result: -639832728, 79` and `Method 1 Concurrent. Arg1 value: 0, 0` are basically random stack values. I only read the values without writing into them. If I start writing, I will get a possibility to easily corrupt the stack.

We could always blame the end consumer for not using the code carefully enough, but when this functionality is hidden behind abstraction layers of NSubstitute/Moq/FakeItEasy/etc, it's very easy to make mistakes. It's especially tricky as e.g. `Method 1 OUTER. Result: 11, 22` still prints the correct result, even though it's use-after-free. It's just stack wasn't touched yet - which make it look like it works. Lastly, not everybody is aware how exactly `ref struct` works, what is stack and what is heap. The idea was to give a convenient abstractions, so most people just use it without any deep knowledge at all.

I think we should make API safer and I have suggestions! :)

### Issue 2

In my example `Method1` and `Method2` are identical, but due to some reason `Method2` is broken when intercepting it. You can see that `Method 2. Result: ` is empty, even though it shouldn't be. Same if I use `out arg`. No matter what I did, I was not able to return the value, even though it works perfectly fine for `Method1`.

### Issue 3
The `AccessViolationException` is too low-level:
```
 System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
```

It's usually thrown by runtime when something weird is happening, or when having errors in P/Invoking native code and escalating access violations back to managed runtime. In our case we are controlling what is happening, so we could just raise just a very specific custom exception and explain what is going on. In all the cases. Either create a custom exception type and provide different messages, or simply throw `InvalidOperationException`. `AccessViolationException` is too scary and misleading, as normally it's for native null-reference exceptions.

## Suggestions

### Suggestion 1

I suggest to use `scoped` keyword to make sure there is no possibility to leak the value at all and use it after it's disposed. When we deference the pointer we could check if it's still alive - so we are safe. Then we also know that consumer has no opportunity at all to store the value somewhere after the method returns.

The keyword was introduced in C# 11/.NET 7,  so it is accessible for all the supported versions.

API could look like this:

<details><summary>`ByRefLikeReference<TByRefLike>`</summary>

```c#
	public unsafe class ByRefLikeReference<TByRefLike> : ByRefLikeReference
		where TByRefLike : struct, allows ref struct
	{
		/// <summary>
		///   Do not use! This constructor should only be called by DynamicProxy internals.
		/// </summary>
		[CLSCompliant(false)]
		[EditorBrowsable(EditorBrowsableState.Never)]
		public ByRefLikeReference(Type type, void* ptr)
			: base(type, ptr)
		{
			if (type != typeof(TByRefLike))
			{
				throw new ArgumentOutOfRangeException(nameof(type));
			}
		}

		private ref TByRefLike Value
		{
			get
			{
				return ref *(TByRefLike*)GetPtrNocheck();
			}
		}

		public TResult UseValue<TResult>(ValueConsumerWithResult<TResult> consumer) where TResult : allows ref struct
		{
			return consumer(Value);
		}
		
		public void UseValue(ValueConsumer consumer)
		{
			consumer(Value);
		}

		public void SetValue(TByRefLike value)
		{
			Value = value;
		}

		public delegate TResult ValueConsumerWithResult<TResult>(scoped TByRefLike value) where TResult : allows ref struct;
		
		public delegate void ValueConsumer(scoped TByRefLike value);
```
</details> 

1. Hide `Value` property and don't expose it directly. Instead introduce `UseValue/SetValue` methods
2. Use `scoped TByRefLike` argument type, so that consumer cannot keep the value somehow.
  Notice, if we use `scoped ref TByRefLike`, the reference itself will be scoped, while the dereferenced value will be not. So we must not use `ref`.
3. Because we cannot use `ref`, we need to use a separate method `SetValue` to update the value.
4. [Optional] Allow to have by ref structs in `TRresult` for convenience. It shall not be a big downside and constrain for normal code, but it could simplify life. I am not totally sure about it.

The usage would be like:

c#
```
        var arg1 = (SpanReference<int>)invocation.Arguments[0]!;
        var returnValue = (SpanReference<int>)invocation.ReturnValue!;

        var arg1Copy = arg1.UseValue((scoped value) => value.ToArray());
        returnValue.SetValue(arg1Copy.AsSpan());
```

Because you cannot leak `scoped value`, you will be forced to make a copy, which introduces total safety.

It of course is a limitation forcing you to make the excessive copies in legit cases as well, but it's the best we could do to guarantee ultimate memory safety. And for me it's 100% worth it.

### Suggestion 2

The suggestion 1 will fix all the possible scenarios for single-threaded code, but it doesn't fix the case of concurrency. As it is possible that background thread could be in the middle of `UseValue` delegate when the original thread will finish the method invocation.

Therefore I suggest to add another simple constrain - allow to access the argument from the original thread only. To me it is a perfectly reasonable constrain, giving that we are dealing with `ref struct` that leave on stack only. Simple thread check will eliminate all the possible concurrency-related bugs and will make it impossible for the reference to escape the scope.

Notice, we'll have to compare the `Thread` object itself, not `ThreadId`, as latter could be re-used (paranoid mode on 😉).

### Suggestion 3

Better protect `ByRefLikeReference.GetPtr()` or remove it all.

I am not totally following how you use it, but I see that you use in the emitted code. The speudo code you provided is still:

<details><summary>Current</summary>

```c#
partial class FooProxy
{
    public unsafe void Method(in Span<byte> bytes)
    {
        Span<byte> bytesCopy = bytes;
        SpanReference<byte> bytesRef = new(..., &bytesCopy); // this is what will go in `IInvocation.Arguments`
        ...
        try
        {
            ... // run through the interception pipeline
        }
        finally
        {
            ...
            bytesRef.Invalidate(&bytesCopy);
        }
    }
```
</details> 

What if instead you do something like:

<details><summary>Suggestion</summary>

```c#
partial class FooProxy
{
    public unsafe void Method(in Span<byte> bytes)
    {
        Span<byte> bytesCopy = bytes;
        ref Span<byte> bytesCopyRef = ref bytesCopy;
        SpanReference<byte> bytesRefArg = new(typeof(Span<byte>), Unsafe.AsPointer(ref bytesCopyRef));
        
        try
        {
            // run through the interception pipeline
        }
        finally
        {
            //...
            bytesRefArg.Invalidate(Unsafe.AsPointer(ref bytesCopyRef));
        }
    }
}
```
</details> 

Then arg object would be just a wrapper over the stack-based `ref` and local code could use `bytesCopyRef` directly without any need to read value back from the arg object. Stack variable will be the source of truth. 

The idea is to eliminate unsafe `ByRefLikeReference.GetPtr()` API to make the feature even more safe.

### Afterword

I believe those suggestions + bug fixes are the must and would be huge enhancement of the safety. Once we do it, I cannot see any possible way to access stack after function exit:
- background threads are prohibited
- we check if wrapper is not disposed before giving the pointer to consumer. So we know we are still on stack
- we know that `ref struct` cannot be leaked because of `scoped` constrain (e.g. to return value or `out/ref` arg)

That will give ultimate safety and will make this feature safe without any technical possibility of misuse. Unless I am missing something :)

**Comment by stakx** [comment]
@zvirja, thank you very much for the time you took testing the new API, as well as your analysis and suggestions. I will need to take some time studying all of it carefully, I just have a few quick responses:

* The reason for your `Method2` test case not working is because you have a copy-paste error in your test code. Fix it as follows:

   ```diff
    private void HandleMethod2(IInvocation invocation)
    {
        var arg1 = (SpanReference<int>)invocation.Arguments[0]!;
   -    var returnValue = (SpanReference<int>)invocation.Arguments[0]!;
   +    var returnValue = (SpanReference<int>)invocation.ReturnValue!;
        ...
   ```

* You are probably right that `AccessViolationException` is not the ideal exception type to throw. Perhaps `InvalidOperationException` would indeed be more appropriate. I'm going to give it some more thought, but will probably change it.

* About making `ByRefLikeReference.GetPtr()` more safe or remove it &ndash; this is tricky. DynamicProxy defines some types and type members that are required by the generated proxies at runtime. `GetPtr` and the pointer-taking ctors are such members. Since the generated proxies are located in a different assembly, these members need to be public lest they become inaccessible to the proxies. Therefore, the best we can do for now is trying to deter users away from such APIs by means of `/// <summary>Do not use!</summary>`, `[EditorBrowsable(EditorBrowsableState.Never)]`, etc. I've considered additionally moving the `ByRefLikeReference` base class into the `Castle.DynamicProxy.Internal` namespace but ended up keeping it in `Castle.DynamicProxy`, along with the other substitute types.

   Once we stop supporting the .NET Framework completely, we can perhaps make these types and type members internal, and emit a `[IgnoresAccessChecksTo]` custom attribute into proxy assemblies instead, so that they still have access to them. But that's not a feasible option just yet.

* I'm still undecided on how to think about your `Method1` test case. I need to study it in more detail and give some thought to how representative it is for actual real-world use cases; but on the surface it reeks of trouble, and I'm thinking that developers accessing another thread's stack frames should expect things to go wrong. Like you're saying: byref-likes aren't the easiest thing to understand. It's a fairly advanced C# language feature, and we (DynamicProxy) can only do so much to avoid trouble for uninformed users. DynamicProxy will probably never be able to guarantee the same safety level as the C# compiler, and I'm not sure it's worth trying... it might be just as important a design goal to keep the API simple.

* That being said: I'll definitely look into your suggestions regarding `scoped`... thanks for that! I must admit that my understanding of this `scoped` is still lacking. Since I've already pushed the `v6.0.0` version tag, and thus cannot go back on the already published API without another major version bump, perhaps we'll add a "safer" API like your `.UseValue` (for more involved use cases) next to the existing `.Value` property (for the more straightforward use cases). But I'll need to give this much more thought.

**Comment by zvirja** [comment]
> The reason for your Method2 test case not working is because you have a copy-paste error in your test code. Fix it as follows...

Oh my god! So stupid. I spent quite some time not understanding at all why it wouldn't work :D 

> About making ByRefLikeReference.GetPtr() more safe or remove it – this is tricky. DynamicProxy defines some types and type members that are required by the generated proxies at runtime

I do understand that! My comment was around a need for the member at all. If we could eliminate the invocation of that method, then the method is not needed. It looks like you are using that method to read the pointer back. But if you just keep `ref` on stack, then you shouldn't have need for that method at all.

At least that was my naive thinking without fully understanding all the code. Disregard if it makes no sense.

> Like you're saying: byref-likes aren't the easiest thing to understand. It's a fairly advanced C# language feature, and we (DynamicProxy) can only do so much to avoid trouble for uninformed users. DynamicProxy will probably never be able to guarantee the same safety level as the C# compiler, and I'm not sure it's worth trying... it might be just as important a design goal to keep the API simple.

My primary worry is not even people not knowing what they are doing - it's mostly the complexity and layers of abstractions. Today while we both are deeply in this code it's obvious how to use it. But then when you have non-trivial code with NSubstitute + AutoFixture + some other libraries doing a ton of magic under the hood - cases could be fairly non-trivial. One thing is that if that leads to a failed test. Or even a randomly failing test. But if that leads to a random stack corruption (especially on some other thread) - that's severe. People will pull their hair trying to understand what is going on 😅 

On a surface it looks that my suggestion would eliminate the issue altogether and there will be no security issues at all. Usage might be slightly cumbersome, but in return you have perfect safety and if code compiles, it means it's safe.

>  Since I've already pushed the v6.0.0 version tag, and thus cannot go back on the already published API without another major version bump, perhaps we'll add a "safer" API like your .UseValue (for more involved use cases) next to the existing .Value property (for the more straightforward use cases). But I'll need to give this much more thought.

Well, technically the package hasn't been published yet due to the technical issues. You could always omit `6.0.0` and publish `6.0.1` with fixed API. Or re-tag `6.0.0`. At the end of the day, release process and SemVer are not the religion - it's also about real life 😉 

Given how easy it is to have use-after-free errors, I suggest to not keep the `.Value`. It's crazy unsafe to me due to a possibility of stack corruption. But that is my opinion.

Sorry for jumping on this a bit late. It's a pity I didn't have a chance to play with it earlier.

Thank you for all the work and time you spend on this!

P.S. Attaching minimal reproduction scenario. It's quite trivial:

<details><summary>Code & Output</summary>
Code:

```c#
using Castle.DynamicProxy;

static class Program
{
    internal static void Main(string[] args)
    {
        var proxyGenerator = new ProxyGenerator();
        var proxy = proxyGenerator.CreateInterfaceProxyWithoutTarget<ISomething>(new PlaygroundInterceptor());

        var result = InvokeMethod(proxy);
        new List<int> { 1, 2, 3 }.Reverse();
        Console.WriteLine($"Method OUTER. Result: {result[0]}, {result[1]}");
    }

    private static Span<int> InvokeMethod(ISomething something)
    {
        Span<int> values = stackalloc int[] { 11, 22 };

        var result = something.Method(values);
        Console.WriteLine($"Method. Result: {result[0]}, {result[1]}");

        return result;
    }
}

public class PlaygroundInterceptor : IInterceptor
{
    public void Intercept(IInvocation invocation)
    {
        var arg1 = (SpanReference<int>)invocation.Arguments[0]!;
        var returnValue = (SpanReference<int>)invocation.ReturnValue!;

        returnValue.Value = arg1.Value;
    }
}

public interface ISomething
{
    Span<int> Method(scoped Span<int> arg1);
}
```

Output:

```c#
Method. Result: 11, 22
Method OUTER. Result: 951577640, 124
```
</details>

**Comment by stakx** [comment]
Thanks for that latest repro. It is a nice demonstration how the current API allows you to casually circumvent C#'s ref safety rules. If an API involving `scoped` can indeed categorically prevent this, then it is definitely worth looking into.

(Another approach would be to recommend interceptors be extra-careful in what they return, especially if they have no knowledge of what the calling code will be doing with the returned byref-likes. For instance, the interceptor could err on the safe side an allocate a copy of the incoming span's data on the heap, then return a span referring to *that*. But I hear you: such usage rule recommendation are much less elegant than being able to actually make the error impossible in the first place using a `scoped`-based API.)

> I spent quite some time not understanding at all why it wouldn't work :D

I was rather shocked seeing you say that even such a simple use case didn't work. If that actually hadn't worked, I would've judged the whole feature an abject and total failure. 😁

**Comment by stakx** [comment]
@zvirja, I've given it some more thought. I am addressing several points, so apologies that this gets a bit long.

---

> #### Suggestion 3
>
> Better protect `ByRefLikeReference.GetPtr()` or remove it all.

I did originally start with an API based solely on managed pointers (`ref`) and `Unsafe.As{Pointer,Ref}` like you suggested [above](https://github.com/castleproject/Core/issues/663#issuecomment-3885803631), instead of using unmanaged pointers (`*`) and all the manual address-taking and pointer dereferencing. Admittedly that would have been much nicer and felt safer, too.

Unfortunately, I soon ran into limitations with that approach, perhaps the major one being that it would have relied on `allows ref struct` anti-constraints in more places, which aren't available on .NET 8. So I also couldn't use `Unsafe.As{Pointer,Ref}` on that platform. Since `net8.0` is still one of this library's supported TFMs, I was looking for other ways to have as-decent-as-possible byref-like support there, too. That's how I ended up with the current unmanaged pointer-based current API.

I am not claiming to have done an exhaustive research over all possible alternative designs, perhaps I missed a better one? But I just didn't see a way forward with something nicely-typed using `Unsafe.As{Pointer,Ref}`.

(As an aside, the current pointer API allows for fairly easy proxy code generation, which is somewhat important with respect to longer-term maintenance of DynamicProxy. Because this API is defined in the base class of the various substitute types `SpanReference<>`, `ReadOnlySpanReference<>` etc., DynamicProxy can treat all of these exactly the same in most places except where they get instantiated; and even there the constructors all have the same signature, so DP only needs to figure out the most derived type to use for a particular parameter. I'm not sure that code generation could have been equally straightforward otherwise.)

---

> #### Suggestion 1
>
> I suggest to use `scoped` keyword to make sure there is no possibility to leak the value at all [...]

As I see it, this suggestion (along with the proposed `UseValue`) has definite merits, but also some potential drawbacks:

* Benefit: It helps prevent ref safety mistakes in certain potentially dangerous situations (such as your recent example code [above](https://github.com/castleproject/Core/issues/663#issuecomment-3886438536)).
* Drawback: It might make it harder or more inefficient to use byref-like arguments in more harmless scenarios; say, when intercepting a method that only has "incoming" byref-like values but does not return byref-likes in any way at all. `UseValue` would force you to wrap parts or all of your actual interceptor code in a callback lambda without any actual benefit. In situations where performance is a concern, this may be undesirable.
* Minor API design issue: Say you intercept a method with two or more byref-like parameters, and you want to access both/several of them simultaneously in your interceptor. You cannot simply nest `UseValue` callbacks inside one another because byref-likes cannot be captured (closed over) and then used "across a lambda boundary". We'd have to provide a whole bunch of `UseValue` overloads for varying numbers of parameters, similar to how there are many variants of the `Action` and `Func` delegate types. This is entirely feasible, but not ideal from a standpoint of API simplicity.
* Drawback: Your `UseValue` proposal appears to be motivated (at least in part) by the idea that making heap-allocated copies of spans is the easy and obvious way to prevent use-after-free errors, and that `UseValue` in combination with `scoped` would force users to make such copies in order to return values. And I partially agree. However, copies may not always be desirable. And not everything is a span. For some byref-like types, there might not be an easy way to copy them to the heap using something as simple as `span.ToArray`. I fear that because of that, your proposed API might end up being too restrictive.

I am almost sure that if we release the current `Value` API into the wild, it won't be long before we get issues of the sort, "Why does this code explode in my face!?", with us having to analyze and explain use-after-free errors and ref safety violations... and I am definitely not looking forward to those.

That being said, I still believe that safety should *not* trump everything else. I guess that's where we fundamentally differ: in the amount of safety DynamicProxy should guarantee. In my eyes, DynamicProxy is a low-level, mostly unopinionated tool with the primary goal of enabling scenarios (here: accessing and manipulating byref-like parameters). Its API *should* be designed in such a way as to steer/guide users towards safe usage, but not necessarily at the cost of preventing legitimate scenarios or efficiency. You're obviously assigning safety a higher priority than I do... which is perfectly fine and understandable.

---

Short intermezzo:

> #### Suggestion 2
>
> [...] Therefore I suggest to add another simple constrain - allow to access the argument from the original thread only.

That's a bit too paranoid for my personal taste. 😁 People accessing stack-only values across threads are basically begging for trouble and I'm not sure its DynamicProxy's responsibility to prevent something so obviously reckless. But yes, the `ByRefLikeReference` base class could perhaps capture the thread and checking it upon access, without taking performance.

Then again, I'm not sure how thread-safe `IInvocation` implementations are today, even without the new byref-like stuff. So perhaps thread safety would have to be considered more widely first, before we start patching single aspects.

---

Where could our different positions regarding the proper amount of safety guarantees meet?

### Option 1: your API

Going with an alternative API identical or similar to the one you suggested, with all of its advantages and drawbacks.

### Option 2: the current API, augmented with your API 

We could add your `UseValue` method as a "safe" choice for people who do not want to educate themselves on the intricacies of ref safety rules, but keep the current API around for "advanced" users.

I was briefly thinking about replacing the read-write `Value` property with a pair of methods, which could then be named `UnsafeGetValue` and/or `UnsafeSetValue`, but such a unsafe hint in the names would be imprecise because neither operation is by itself inherently unsafe; the unsafety results from how the methods are combined. So replacing or renaming the property would probably not have much benefit.

### Option 3: Better documentation

The [current documentation](https://github.com/castleproject/Core/blob/a92fe781ffce10301227a4ae0e5b438aa926992d/docs/dynamicproxy-byref-like-parameters.md) completely overlooks dangerous scenarios where things can go wrong, like yours shown above. The current usage recommendations aren't sufficient, I think the documentation needs additional information about ref safety.

Let's quickly go back to your latest code example. I've noted that the method you're intercepting has a `scoped` parameter:

```csharp
public interface ISomething
{
    Span<int> Method(scoped Span<int> arg1);
}
```

If it weren't for that `scoped`, the C# compiler would have rejected your `InvokeMethod` implementation due to a potential ref safety violation. I don't have a perfect intuitive understanding of `scoped` yet, but here, `scoped` can be understood as a "promise" that `Method` will not let `arg1` "escape" from it.

I think it would be fair to expect interceptors to heed that promise, too. Perhaps we add a paragraph or two about interceptors having to be careful what they are returning, especially when the intercepted method promises not to leak `scoped` parameters; and that interceptors should make defensive data copies if in doubt...?

### Option 4: Leave it up to downstream code (FakeItEasy, NSubstitute, etc.) to prevent unsafe scenarios

I think it's safe to say that most developers only ever come into contact with DynamicProxy indirectly, via the popular mocking libraries. Typically, these libraries have an API of their own and typically don't expose DynamicProxy's directly to the user. Therefore I think it's fair to expect that when it comes to byref-likes and ref safety, these libraries should share some of the responsibility for providing safety. For example, it would be entirely possible for a mocking library to perform span copying by default, and never expose `SpanReference<>` etc. at all (at the cost of perhaps having no support for byref-like types other than spans).

### Option 5: scratch the current API

... and do something totally different, or nothing at all.

---

I personally am gravitating towards both options 3 (documentation) which I think is a must, and 4 (let mocking libraries provide additional safeties) which I think is a fair ask and might be a fitting split of responsibilities at different levels.
 
I'd be OK with 2 (my and your APIs as side-by-side alternatives). I think I am against 1 (just your API) as it seems to restrictive to me (see concerns outlined above), and I am definitely not very keen on 5 (start over again) but it's an option that someone else might want to look into.

**Comment by zvirja** [comment]
Hey @stakx!

Thanks for the verbose and detailed analysis of my reply and for the possible options. I see your reasoning and would like to elaborate on it.

## Safety

You are totally right to see that we have a very different view on the concern around safety and for me it's indeed a top priority. Unlike multi-threaded programming issues (like race-conditions), stack corruption is way more dangerous and is a totally different class of problems. If you mess up with stack, you can make stack unwalkable (if you somehow corrupt the return address), lead to peculiar GC behavior - not even mentioning bizzare computation results. The critical issue here is that runtime is not well-prepared for such kind of problems. Neither are most of the developers, even senior. It's not just a regular `NullReferenceException` which is handled pretty well - runtime could just start hallucinating. It's not a joke at all. Especially because it could happen very randomly, as you never know what people above a few levels of abstractions are doing. People who use it via some other third-party libraries might not have even a slightest idea what DynamicProxies and Interceptors are. The cost of troubleshooting it and frustration it could bring is just beyond imagination.

Another aspect is that current API would violate .NET language philosophy in my opinion. Currently the runtime promise is that unless you mess up with `unsafe` or `P/Invoke`, it is impossible to corrupt CLR no matter what you do. Here we are offering very much safe API which gives you means to do it. You can break CLR easily staring to deal with `Unsafe.*` or pointers - but then you have to be in the `unsafe` context, so it's kind of obvious that you are a bit on your own.

It would be careless to provide such a dangerous API so casually. We cannot just say "just read the documentation, we warned you" as again:
- nobody realistically reads the documentation in most of the cases - people just use the API available
- people might use code indirectly via other libraries
- code could be non-trivial, so it might be very tricky to spot the bug even if you know what you are doing
  - code could evolve by refactorings, so it might be invisible that you introduce an issue

Offloading responsibility on the mock library developers also doesn't look good or realistic:
- if underlying abstraction is unsafe, the exposed API could be similarly unsafe. It could be transient.
- it might be not realistically possible to do analysis of what user is doing with the API to verify it's legit. Like I am a bit aware of `NSubstitute` code and I cannot imagine doing proper analysis to detect that users assign arguments to return values in a wrong way
- if all the libraries have to build safety around unsafe API and API is dangerous to be used as-is, then it might be a good idea to provide better API in the first place.

The basic API shall be totally safe, even if it forces you to have overheads and inconveniences. We could provide means for optimizations, but that API shall not be the primary one if it's unsafe.

> In my eyes, DynamicProxy is a low-level, mostly unopinionated tool with the primary goal of enabling scenarios (here: accessing and manipulating byref-like parameters). Its API should be designed in such a way as to steer/guide users towards safe usage, but not necessarily at the cost of preventing legitimate scenarios or efficiency.

We come a bit into design philosophy, but speaking generically would say that if a certain feature has a risk of stack corruption by using it wrongly, then I would probably go as far as saying that we shall refuse to admit that the feature is possible at all. Because runtime safety is not something you negotiate and take optionally - it's a hard constrain. You could for example create a library which allows you to hot-swap two methods my messing up with MethodTable structures - so when function is invoking one method, it invokes a substitute implementation instead. It's a powerful feature - you can use it for diagnostics, to log third-party methods and so on. Very handy. But the risks are above the roof and what about inlining, tiered-compilation and so on. I've tired doing it myself, it works and code is simpler than the one you wrote here for this feature. But just because it is doable - it's not enough to say that it's something to consider even as an option.

DynamicProxy is indeed a low-level tool. But so far you never had a risk of corrupting runtime by using some features without enough care. The biggest risk was a weird exception when you used a language feature which wasn't supported yet. So it would be a quality change in the nature of the provided API.

Okay, I feel I am boring you at this point :) I guess you got my point that IMO it's not just a risk - it's a critical risk.

## Thread Safety

Checking for thread costs nothing, but it will prevent weird behavior and stack corruption. Of course it's stupid to access it from another thread. But again - you never know what happens above the layers of abstractions and how complex the code is. 

Stack corruption and concurrency issues - are very different issues. So we cannot use non-thread-safety of `IInterceptor` API as an argument - it's a totally different class of the issues.

It's like .NET implements finalizers for handles wrappers. You shall in principle dispose things you own. But they would rather play safe to not leak native resources. Even though they mention in the documentation that you shall dispose things. And there is a compiler warning (if I am not mistaken).

## Inefficiency

> Minor API design issue: Say you intercept a method with two or more byref-like parameters, and you want to access both/several of them simultaneously in your interceptor.

> ... We'd have to provide a whole bunch of UseValue overloads for varying numbers of parameters, similar to how there are many variants of the Action and Func delegate types. This is entirely feasible, but not ideal from a standpoint of API simplicity. ...

> However, copies may not always be desirable. And not everything is a span. For some byref-like types, there might not be an easy way to copy them to the heap using something as simple as span.ToArray. I fear that because of that, your proposed API might end up being too restrictive.

So far we are building the first version and enabling the scenario which was not possible at all. We don't know how API is going to be used and whether it would be a real concern. People might be totally OK with it if they are explained that otherwise there could be a huge risk of runtime integrity issues. A lot of code is just for testing and there it's totally OK to have not the most effective code.

Some things like accessing multiple arguments at the same time might be not a real issue due to the way how mock libraries expose the API on their level. It's hard to guess.

I wouldn't be worried about it ahead of time. I would just provide the first version of API and look for the feedback. If such a concern comes, we could always add more API and think more about it later.

I am very much a performance guy myself, so I perfectly relate to you and see where you are coming from. It's just hard to understand if the concern is real. Consumers might just prefer safety. But nevertheless I still have an idea how to satisfy it.

## Options feedback

### Option 3: Better documentation

It is not a solution at all in my opinion. We cannot rely that realistically people will read it. And those who read might not fully understand all the cases it concerns. Sure, maybe AI will. But it's just more like an apology to me rather than something real.

### Option 4: Leave it up to downstream code (FakeItEasy, NSubstitute, etc.) to prevent unsafe scenarios

Be kind to your fellow brothers (yourself included) :D Now we rely that 3+ mocking libraries properly analyze the domain, understand all the implications and scenarios (I am not even sure about myself; I found a couple of simple repro cases, but I am pretty sure there are more exotic ones) - and correctly handle the concerns. Each has to take the responsibility (without maybe even fully realizing it) that wrong implementation could lead to stack corruption and C-world fun in managed runtime. I would rather not smear the problem everywhere, as then chances of making it wrong approach 100%.

## Suggested solution

I suggest slightly modified hybrid of Option 1 and Option 2.

### Option 1

Just go with providing safe (and limiting) API only and wait for the feedback. If there would be feedback that API is not enough - then look at the scenarios and add extra unsafe API.

### Option 2

This option is a nice of compromise which shall cover your concerns. Let's follow the design of the [`CollectionsMarshal`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal?view=net-10.0). The idea is that main collections don't expose unsafe API, but you can use some other class to have features with higher risk.

- I suggest to only have `UseValue/SetValue` API on the wrappers
- Introduce `ByRefLikeReferenceUnsafe/ByRefLikeReferenceMarshal` which allows to read `ref struct` without `scoped` protection:
   - still keep Dispose protection
   - still keep the origin thread protection
   - make methods/class `unsafe` to require usage of `unsafe` context.

This way:
- You cannot corrupt the stack if use just primary API
- If you need optimizations, you still could do it. But then you enter the `unsafe` code territory
- You explicitly enter `unsafe` code which is an extra guard that you are doing something risky
- The API exists "on a side", so it's not immediately discoverable. You only look for it if you need it (or read the documentation) and then you probably know what you are doing

I would say that code must be unsafe when you have unprotected ref struct. Just like pointers - there is nothing wrong with using them. But you can easily shoot yourself in a leg if you overlook something. Same here. It's mostly about risks, not about what you do per se.

I think it's a beautiful compromise. As long as you use safe API, we promise that there are no issues in a price of inconveniences. There is unsafe API on a side to enable writing more efficient code, but it imposes risks and requires you to be more careful, just like `Unsafe.*`.

### Option 3. Bonus

Apparently we already could use `Unsafe` API to bypass `scoped` protection (as long as you add `allow ref struct` anti-constrain to `UseValue`) 😳😅

```c#
var arg1Unprotected = arg1.UseValue((scoped value) => Unsafe.AsRef<Span<int>>(Unsafe.AsPointer(ref value)));

var arg1Unprotected = arg1.UseValue((scoped value) => Unsafe.Read<Span<int>>(Unsafe.AsPointer(ref value)));

var arg1Unprotected = arg1.UseValue((scoped value) => Unsafe.As<Span<int>, Span<int>>(ref value));
```

Maybe there is something else. All of it nicely bypasses the `scoped` keyword, so you are free to do what you want afterwards. It looks a bit scary - but it is risky to do it, so it is a "feature" in a way.

The fact that such API exists and is easily available advocates to go with Option 1:
- there is a performance/convenience workaround
- we could always add API for Option 2 if we find it's justified

But I do see that it's not very elegant and a bit cumbersome. It's just that it's possible. Marshal API would look nicer than this of course.

## P.S. Better protect ByRefLikeReference.GetPtr() or remove it all.

I feel like we have a bit of misunderstanding. Or I have. Could you explain how exactly you use `.GetPtr()` API? My comment was not about class hierarchy or `ByRefLikeReference` type itself - it was only about that single specific method. My point was that if you keep data on stack, then it looks like you don't need that API at all. Because then you could simply look at the data on stack.

Your reply has a different gist (or maybe I read it this way), so let's align first.

## Thanks

Thanks for your time and patience @stakx! It's a nice discussion and I have a lot of fun 😊 

It's very late and I spent around 2.5 hours writing this. I apologize in advance if it's a bit repetitive - I don't have more energy to tidy it up.

**Comment by stakx** [comment]
@zvirja, thanks for the time and thoughts you're putting into this, I appreciate it. I'm going to answer more fully later, right now I'm a little sleepy, so just something quick.

Suppose we do it your way and change the API of e. g. `SpanReference<T>` as follows:

> ```diff
> -public ref Span<T> Value
> +private ref Span<T> Value
>  {
>      get
>      {
>          return ref *(Span<T>*)GetPtrNocheck();
>      }
>  }
> +
> +public TResult UseValue<TResult>(ValueConsumer<TResult> consumer)
> +{
> +    return consumer(Value);
> +}
> +
> +public void UseValue(ValueConsumer consumer)
> +{
> +    consumer(Value);
> +}
> +
> +public void SetValue(Span<T> value)
> +{
> +    Value = value;
> +}
> +
> +public delegate void ValueConsumer(scoped Span<T> value);
> +
> +public delegate TResult ValueConsumer<TResult>(scoped Span<T> value);
> ```

What are we going to do about the following?:

```csharp
public interface IFoo
{
    Span<char> Method();
}

public class BadInterceptor : IInterceptor
{
    public void Intercept(IInvocation invocation)
    {
        Span<char> chars = stackalloc char[10];
        "Hi".CopyTo(chars);

        var returnValueRef = (SpanReference<char>)invocation.ReturnValue!;
        returnValueRef.SetValue(chars);
    }
}
```

Or, a little more involved, but closer to your previous example:

```csharp
Span<char> InvokeMethod()
{
    Span<char> chars = stackalloc char[10];
    "Hi".CopyTo(chars);

    var generator = new ProxyGenerator();
    var foo = generator.CreateInterfaceProxyWithoutTarget<IFoo>(new BadInterceptor());
    return foo.Method(chars);
}

public interface IFoo
{
    Span<char> Method(scoped Span<char> arg);
}

public class BadInterceptor : IInterceptor
{
    public void Intercept(IInvocation invocation)
    {
        var argRef = (SpanReference<char>)invocation.Arguments[0]!;
        var returnValueRef = (SpanReference<char>)invocation.ReturnValue!;
        argRef.UseValue((scoped Span<char> arg) =>
        {
            returnValueRef.SetValue(arg);
        });
    }
}
```

**Comment by zvirja** [comment]
@stakx Amazing cases!!! 🤯🤩 It changes a lot my mental model of the problem and are nice examples indeed that the API I suggested are not fully safe either.

Let me play with it for a tiny bit. So far I have 2 workable ideas, but I am curious to see if I could find something more elegant :)

**Comment by stakx** [comment]
@zvirja, I propose a new compromise: we initially get rid of any means to set outbound byref-like-typed parameters. (DynamicProxy will write `default(TByRef)` where necessary.) A value getter on its own should not be able to cause any stack corruption, and the checks/invalidations that are already in place should sufficently guarantee that one cannot read byref-like parameters that have gone out of scope.

Specifically, I propose to replace the `ref TByRefLike Value { get; }` property with a `TByRefLike GetValue()` method. Note the absence of `ref` in the return type, and the absence of a setter method.

Then we wait and see if this read-only solution is actually sufficient for most real-world use cases.

* If so: great, we no longer have a safety problem!
* If not: only then do we consider adding back a setter, but we would call it `void UnsafeSetValue(TByRefLike)` to draw very obvious attention to the potential unsafety, and link it to more detailed documentation about the risks (including concrete examples) of using this method incorrectly.

Regarding other, less pressing points:

* I think we should indeed throw `InvalidOperationException` instead of `AccessViolationException`. I'll make the change.
* I don't think checking for cross-thread access is truly necessary esp. when it isn't possible to set byref-like parameters; the existing safety measures should already suffice.

**Comment by zvirja** [comment]
Good evening, @stakx! Sorry for a bit belated reply. I took a bit of time to research on this, to play with it and to let it settle inside myself to see if I come up with something better. Nothing better comes, so I'll just share what I found and my thoughts. Let's see if we could advance the story together.

I would like to immediately say that this story quickly becomes messy, as there is no elegant solution. There are good enough solutions, even though they are a bit weird at the first glance. Part of me feels a bit sad that I didn't manage to come up with something extremely neat. I feel a bit of pressure, as I feel that I am a bit alone on this "safety" side and at the moments of despair I start to have doubts as well 😅  But I still believe that we shall not expose things which could lead to stack corruption and as long as it's technically possible to do that, I would suggest to stick to that even if API is a bit weird.

I do like a lot your approach of phased API exposure. This way we indeed could ship basic version of the feature and only care about more advanced scenarios if the need comes. I could easily see it happening that we might need to reshape API anyway as more people would start consuming it.

I would first start with answering to your points and later will describe my analysis and ideas. Bear with me. Writing this reply was quite a journey :)

# Feedback on your proposals

## `TByRefLike GetValue()`-only method

It took me a bit of time to assess if this API is actually safe. My initial idea of breaking it would be to emulate something like:

```c#
    public ref struct Container
    {
        public ref int Value;
    }
    
    public interface ISomething
    {
        void Method(scoped Span<int> arg1, ref Container arg2)
        {
            arg2.Value = ref arg1[0];
        }
    }
```

It however is not possible, as you will never return original `arg2` instance, only a copy of the struct arg. If there is no other way how we could modify the original argument and carry the value out, then it shall be safe indeed.

But it is only safe if we never return values. If we want to support returning values, then API causes troubles and doesn't allow to build safe API. I'll elaborate on it below.

I personally think that it's a big limitation, as then basically it's impossible to set `ref struct` return values and those are used quite a lot. I would suggest to explore writing API as well. I think it would be fine if it supports basic scenarios only - but it shall be possible to return the value. It will also allow mocking libraries to create the full feature.

## Thread-safety

> I don't think checking for cross-thread access is truly necessary esp. when it isn't possible to set byref-like parameters; the existing safety measures should already suffice.

It is, as we could then have scenario like this:

<details><summary>Code</summary>

```c#
using Castle.DynamicProxy;

namespace PointersPlayground;

public static class Program
{
    private class HandlerInterceptor(Action<IInvocation> handler) : IInterceptor
    {
        public void Intercept(IInvocation invocation) => handler(invocation);
    }

    public interface ISomething
    {
        void Method(Span<int> arg1);
    }
    
    internal static void Main(string[] args)
    {
        var mre1 = new ManualResetEventSlim(false);
        var mre2 = new ManualResetEventSlim(false);
        var mre3 = new ManualResetEventSlim(false);
        
        var proxyGenerator = new ProxyGenerator();
        var proxy = proxyGenerator.CreateInterfaceProxyWithoutTarget<ISomething>(new HandlerInterceptor(invocation =>
        {
            var arg1 = (SpanReference<int>)invocation.Arguments[0]!;

            Task.Run(() =>
            {
                var arg1ValueCopy = arg1.GetValue();
                
                mre1.Set();
                mre2.Wait();

                Console.WriteLine($"Concurrent Arg 1: {arg1ValueCopy[0]}");
                
                mre3.Set();
            });
            
            
            Console.WriteLine($"Arg 1: {arg1.GetValue()[0]}");

            mre1.Wait();
        }));

        InvokeMethod(proxy);

        Span<int> data = stackalloc[] { 1, 5, 5, 7, 8, 9, 5, 44, 3, 2 };

        mre2.Set();
        mre3.Wait();
    }

    private static void InvokeMethod(ISomething proxy)
    {
        Span<int> arg1 = stackalloc[] { 42 };
        
        proxy.Method(arg1);
    }

    private static TByRef GetValue<TByRef>(this ByRefLikeReference<TByRef> value) where TByRef : struct, allows ref struct => value.Value;
}
```
</details> 

Output:
```
Arg 1: 42
Concurrent Arg 1: -1002968512
```

# My analysis

## `GetValue()` scoped leakage

You gave me a truly amazing scenario with leaking stack-allocated data and it changed a bit the way how I was thinking about the problem. I started to consider whether the issue is with returns only, or whether there is another way of misusing and leaking the argument if it's not protected by `scoped` constrain.

I found the following sources of "leaking" the scoped value:
- `return`
- `ref/out`
- interior mutability of another ref-struct argument

I haven't realized that we in fact could return argument value copies everywhere, even if it's `ref Span<int>`. If so, then leaking through the interior mutability shall not be possible (because `RefStruct` cannot have `ref RefStruct` field). It took me some time to try to break it with the analysis above and explore it. I haven't found a way. That in theory significantly simplifies the issue, as then the problem is narrowed to not leaking scoped values via `SetValue()` (as that API is used both for `return` and for `ref/out`).

If we solve the `SetValue()` problem, then we could return scoped arguments unprotected (i.e. without `scoped` constrain) - there is no way to leak them otherwise. The issue, unfortunately, is that `SetValue()` now becomes trickier, as it shall protect not only the `scoped ref struct`, but basically any `ref struct` - while you still want to somehow craft the result. I was not able to find any implementation that would work.

## SetValue() scoped leakage

We need to make sure that `scoped` variables are not leaked. Like:
- prevent leaking local `stackalloc` variables
- prevent leaking `scoped` arguments

### Lambda-based approach

My initial intuition was to leverage the limitation you already spotted above that `ref struct` cannot be captured in lambda. We also `return` the value there, so compiler shall protect the `scoped` variable. Then the solution would be quite nice and neat:

```c#
	public unsafe class ByRefLikeReference<TByRefLike>
	{
		public void SetValue(ValueGetter valueGetter)
		{
			Value = valueGetter.Invoke();
		}
		
		// nit: Func<> might work as well - I was playing with API :)
		public delegate TByRefLike ValueGetter();
	}

var returnArg = (SpanReference<int>)invocation.ReturnValue!;
returnArg.SetValue(() => new[] { 42 });
```

Yes, it might look a bit cumbersome - but then it's totally safe if all the scoped arguments are `scoped` protected. We basically return back the compiler check and could trust it doing its job.

The issue comes if we have `TByRefLike GetValue()` API which strips `scoped` modifier from the argument. As then it's totally possible to write the following and we are back to square one:

```c#
        invocation =>
        {
            var arg1 = (SpanReference<int>)invocation.Arguments[0]!;
            var returnArg = (SpanReference<int>)invocation.ReturnValue!;

            returnArg.SetValue(() => arg1.GetValue());
        }
```

So this solution is totally safe only if we have scoped args protected.

### Idea - ref struct based approach

Another idea I explored was API like `arg1.ValueAcessor.SetValue(TByRefLike value)`. If `ValueAccessor` is `ref struct`, then compiler would not allow you to give it scoped value.

While it works, I am not sure how strong the guarantee is in this case. To me it sounds like it currently works because of flow analysis limitation and in the future they could enhance it. As in principle compiler could conclude that `ValueAccessor` cannot possibly outlive the current stack frame, so it's safe to accept scoped value.

My knowledge is not strong enough to see it fully. Given that API is not elegant anyway, I decided that lambda-based approach is better. 

Notice, this approach also provides total safety only if we don't strip `scoped` protection from the argument.

## Scoped argument awareness

Given that it is possible to have totally safe `SetValue()` only if `scoped` is respected, we need to preserve it and cannot strip it. I was considering what are the options.

Simple one is overprotection. We could provide `UseValue()` API only, so all the values are `scoped`-protected. It's overkill for 99% of the cases, but it's safe.

Second option would be to introduce library awareness of scoped arguments. Given that `scoped` arguments are used rarely, we could then provide more convenient API for the regular arguments without trading-off safety.

First, we shall extend parent class with a flag

```c#
public unsafe class ByRefLikeReference
{
	public bool IsScoped { get; }
}
```

Then it's fairly easy to populate the value in the dynamic proxy:

```c#
// be aware of `scoped ref Span<int>` - value itself is not scoped in that case
var isScoped = parameters[i].GetCustomAttribute<ScopedRefAttribute>() != null && !parameters[i].IsByRef;
method.CodeBuilder.AddStatement(
	new AssignStatement(
		reference,
		new NewInstanceExpression(
			referenceCtor,
			new TypeTokenExpression(dereferencedArgumentType),
			new AddressOfExpression(dereferencedArgument),
			new LiteralBoolExpression(isScoped))));
```

For the return values it's always non-scoped.

If we have this awareness, then we could have API like this:

```c#
public void UseValue(ValueConsumer consumer)
{
	consumer(Value);
}

public TResult UseValue<TResult>(ValueConsumerWithResult<TResult> consumer) where TResult : allows ref struct
{
	return consumer(Value);
}

public TByRefLike GetValue()
{
	if (this.IsScoped)
	{
		throw new InvalidOperationException("Use UseValue method for scoped arguments");
	}

	return Value;
}

public void SetValue(ValueGetter valueGetter)
{
	Value = valueGetter.Invoke();
}
```

For 99% of the cases we could just use `arg.GetValue()` without paying price of `UseValue()` limitations. If you have scoped argument - then you would be forced to use the method.

## Suggested solution

I suggest to:
- Add thread-safety checks
- Introduce library awareness around `scoped` arguments - it's quite trivial
- Provide `GetValue()/UseValue()` with constrain that for scoped arguments it's only possible to use `UseValue()`
- Provide lambda-based `SetValue()` API
- [optional] Provide `ByRefLikeReferenceUnsafe` unsafe API to access scoped arguments/set arbitrary value without protection. I would hide it from direct API. Maybe not add it at all, unless we find a need for it.
    - If you introduce this type, we could then move there all the unsafe API instead of just hiding it from intellisense like you do it now. Just make it internal and expose via `ByRefLikeReferenceUnsafe`:
    	- `void* GetPtr(Type checkType)`
    	- `Invalidate(void* checkPtr)`
    	- contructors
      This way API will be cleaner and you will have better control over activation

Unless I am missing something, it shall enable all the possible scenarios with total safety:
- You can return heap-based ref struct value
- You can return another argument (for `scoped` `arg` it will throw):
    ```c#
        invocation =>
        {
            var arg1 = (SpanReference<int>)invocation.Arguments[0]!;
            var returnArg = (SpanReference<int>)invocation.ReturnValue!;
            
            returnArg.SetValue(() => arg1.GetValue());
        })
    ```
- You cannot return local variables
- You cannot return scoped arguments

I like that API is safe, quite trivial (except unusual `SetValue()`) and for non-scoped arguments it's extremely convenient. If there are some limitations, we could address them in V2. But so far it looks like all the scenarios are covered.

Another big pros is that `scoped` will never be stripped, so we will have proper compiler check for all other possible leakage scenarios we haven't considered.

What do you think?

P.S. After doing this analysis I am even more confident that we shall ship safe API only. Analyzing it all was quite a hell and one cannot expect normal human to be aware of all of this 🤯 If API is dangerous, one will definitely shoot those legs one day or another.

**Comment by stakx** [comment]
Apologies for the delay @zvirja. I'll see if I can look into this on the weekend and reply to your latest ideas and suggestions then.

**Comment by zvirja** [comment]
@stakx No worries! I saw your idea to release it as soon as possible and I do believe that we shall incorporate safety for `ref struct` in the very first release. That's better than deprecating it all later and changing the surface. I would be more than happy to prepare PR myself, so you could later do the follow up. That shall reduce the pressure from you and distribute the load a bit 😉 

I'll try to look in the coming day/s and build it.

**Comment by stakx** [comment]
I did say "as soon as possible" but I don't intend to completely sacrifice this ongoing discussion for that, so don't worry. On the other hand, I do hope we'll soon reach a point where we can ship something _reasonably_ safe. I'd also prefer shipping an incomplete solution rather than one we'll have to go back on.

(I still don't believe that 100% ref safety will be achievable btw.)

**Comment by zvirja** [comment]
@stakx I got a bit busy these days, but would be happy to work on it on Monday if you need some time to enjoy spring. Would feel happy to help!

**Comment by stakx** [comment]
@zvirja, I'm a little late too but let me take a more detailed look at your post above first, I'll get back to you.

**Comment by drauch** [comment]
Is there still ongoing work on this topic or is future ref support now gone? 

Best regards,
D.R.

**Comment by stakx** [comment]
No, not gone, I just need to find a quiet moment to finalize the pending ref safety concerns. I haven't forgotten about it, I'm just a bit short on time.

An approach to address ref safety was suggested above though it seems somewhat too complex IMHO, and I'd prefer to keep the API as simple as possible. For me, the likelier way to address the safety concerns would be to scale back the current byref-like support to a pure read-only interface, for starters... and only add back write support if and when people actually demand it. Admittedly, that only delays having to deal with the ref safety Pandora's box to some point in the future, but that's probably still better than not releasing anything at all.

Apologies for the wait, in either case.
